# Running `phold`

`phold` has been split into two overall modules regardless of your input - `predict` and `compare`. 

The reason `phold` has been split as such is to enable optimal resource usage between GPU and CPU in cluster environments.

If you have a local workstation with an NVIDIA GPU, please use `phold run`, which combines both in one.

If you have a local workstation without a GPU, please use `phold run` with `--cpu`.

If you have trouble running ProstT5, please use `phold remote`.

If you are having troubing runing `phold` offline after installation (e.g. on a HPC), please add `TRANSFORMERS_OFFLINE=True` to your environment. 

## Input 

Most subcommands of `phold` takes as their input an entry Genbank formatted file that contains the output of [`pharokka`](https://github.com/gbouras13/pharokka) for your phage or phage contigs. This will be called `pharokka.gbk` by default in your pharokka output.

Alternatively, `phold` will detect if the input is a FASTA contig/genome file as input. If so, [Pyrodigal-gv](https://github.com/althonos/pyrodigal-gv]) will then be run to quickly predict the CDS and these will be annotated. However, neither tRNAs, tmRNA nor CRISPR repeats will be predicted (unlike in pharokka) for now.

For `phold proteins-predict` and `phold proteins-compare`, the input will be a FASTA format file containing amino acid protein sequences. These commands are useful for annotating bulk phage proteins. 

## Subcommands

### `phold predict`

`predict` uses the [ProstT5](https://github.com/mheinzinger/ProstT5) protein language model to translate protein amino acid sequences to the 3Di token alphabet used by [foldseek](https://github.com/steineggerlab/foldseek). This module is greatly accelerated if you have a GPU available and is recommended

```bash
Usage: phold predict [OPTIONS]

  Uses ProstT5 to predict 3Di tokens - GPU recommended

Options:
  -h, --help                     Show this message and exit.
  -V, --version                  Show the version and exit.
  -i, --input PATH               Path to input file in Genbank format or
                                 nucleotide FASTA format  [required]
  -o, --output PATH              Output directory   [default: output_phold]
  -t, --threads INTEGER          Number of threads  [default: 1]
  -p, --prefix TEXT              Prefix for output files  [default: phold]
  -d, --database TEXT            Specific path to installed phold database
  -f, --force                    Force overwrites the output directory
  --batch_size INTEGER           batch size for ProstT5. 1 is usually fastest.
                                 [default: 1]
  --cpu                          Use cpus only.
  --omit_probs                   Do not output 3Di probabilities from ProstT5
  --finetune                     Use finetuned ProstT5 model (PhrostT5).
                                 Experimental and not recommended for now
  --finetune_path TEXT           Path to finetuned model weights
  --save_per_residue_embeddings  Save the ProstT5 embeddings per resuide in a
                                 h5 file
  --save_per_protein_embeddings  Save the ProstT5 embeddings as means per
                                 protein in a h5 file
```

Example usage (assuming you have run `phold install`)

```bash
phold predict -i pharokka.gbk -o phold_predict_output 
```

### `phold compare`

`phold compare` runs [foldseek](https://github.com/steineggerlab/foldseek) to compare ProstT5 predictions generated by `phold predict` to the phold database.

Alternatively, if you have provided pre-generated .pdb format protein structures for you proteins, you can specify those by specifiying `--structures --structure_dir <directory>`. 

`phold compare` does not use a GPU. Use as many CPU threads with `-t` as you can.

Example usage following `phold predict`

```bash
phold compare -i pharokka.gbk -o phold_compare_output  --predictions_dir phold_predict_output
```

Example usage if you have .pdb format structures available for your phage proteins (note: the .pdb file names must be called cds_id.pdb, where cds_id is the CDS ids output from Pharokka). You can see an example in the `tests/test_data/NC_043029_pdbs` directory [here](https://github.com/gbouras13/phold/tree/main/tests/test_data/NC_043029_pdbs).


```bash
phold compare -i pharokka.gbk -o phold_compare_output_pdb  --pdb --pdb_dir directory_with_pdbs  -t 8
```

```bash
Usage: phold compare [OPTIONS]

  Runs Foldseek vs phold db
Options:
  -h, --help                Show this message and exit.
  -V, --version             Show the version and exit.
  -i, --input PATH          Path to input file in Genbank format or nucleotide
                            FASTA format  [required]
  --predictions_dir PATH    Path to output directory from phold predict
  --structures              Use if you have .pdb or .cif file structures for
                            the input proteins (e.g. with AF2/Colabfold .pdb
                            or AF3 for .cif) in a directory that you specify
                            with --structure_dir
  --structure_dir PATH      Path to directory with .pdb or .cif file
                            structures. The CDS IDs need to be in the name of
                            the file
  --filter_structures       Flag that creates a copy of the .pdb or .cif files
                            structures with matching record IDs found in the
                            input GenBank file. Helpful if you have a
                            directory with lots of .pdb files and want to
                            annotate only e.g. 1 phage.
  -o, --output PATH         Output directory   [default: output_phold]
  -t, --threads INTEGER     Number of threads  [default: 1]
  -p, --prefix TEXT         Prefix for output files  [default: phold]
  -d, --database TEXT       Specific path to installed phold database
  -f, --force               Force overwrites the output directory
  -e, --evalue FLOAT        Evalue threshold for Foldseek  [default: 1e-3]
  -s, --sensitivity FLOAT   Sensitivity parameter for foldseek  [default: 9.5]
  --keep_tmp_files          Keep temporary intermediate files, particularly
                            the large foldseek_results.tsv of all Foldseek
                            hits
  --card_vfdb_evalue FLOAT  Stricter Evalue threshold for Foldseek CARD and
                            VFDB hits  [default: 1e-10]
  --separate                Output separate GenBank files for each contig
  --max_seqs INTEGER        Maximum results per query sequence allowed to pass
                            the prefilter. You may want to reduce this to save
                            disk space for enormous datasets  [default: 10000]
  --only_representatives    Foldseek search only against the cluster
                            representatives (i.e. turn off --cluster-search 1
                            Foldseek parameter)
  --ultra_sensitive         Runs phold with maximum sensitivity by skipping
                            Foldseek prefilter. Not recommended for large
                            datasets.
```

### `phold run`

`phold run` runs `phold predict` and `phold compare` together in one command. Recommended if you are running `phold` on a local workstation with an available GPU.

Also recommended if you are running `phold` in a low-resource environment without a GPU e.g. with a laptop - you will also need to specify `--cpu`.

Example usage where GPU is available:

```bash
phold run -i pharokka.gbk -o phold_output  -t 8
```

```bash
Usage: phold run [OPTIONS]

  phold predict then comapare all in one - GPU recommended

Options:
  -h, --help                Show this message and exit.
  -V, --version             Show the version and exit.
  -i, --input PATH          Path to input file in Pharokka Genbank format or
                            nucleotide FASTA format  [required]
  -o, --output PATH         Output directory   [default: output_phold]
  -t, --threads INTEGER     Number of threads  [default: 1]
  -p, --prefix TEXT         Prefix for output files  [default: phold]
  -d, --database TEXT       Specific path to installed phold database
  -f, --force               Force overwrites the output directory
  --batch_size INTEGER      batch size for ProstT5  [default: 1]
  --cpu                     Use cpus only with ProstT5
  --omit_probs              Do not output 3Di probabilities from ProstT5
  --finetune                Use finetuned ProstT5 model
  --finetune_path TEXT      Path to finetuned model weights
  -e, --evalue FLOAT        Evalue threshold for Foldseek  [default: 1e-3]
  -s, --sensitivity FLOAT   sensitivity parameter for foldseek  [default: 9.5]
  --card_vfdb_evalue FLOAT  Stricter Evalue threshold for Foldseek CARD and
                            VFDB hits  [default: 1e-10]
  --separate                output separate genbank files for each contig
  --keep_tmp_files          Keep temporary intermediate files, particularly
                            the large foldseek_results.tsv of all Foldseek
                            hits
  --split                   Split the Foldseek runs by ProstT5 probability
  --split_threshold FLOAT   ProstT5 Probability to split by  [default: 60]
  --max_seqs INTEGER        Maximum results per query sequence allowed to pass
                            the prefilter. You may want to reduce this to save
                            disk space for enormous datasets  [default: 1000]
```

### `phold proteins-predict`

Identical to `phold predict`, but instead takes a FASTA input file of amino acid protein sequences. Useful for bulk annotation of phage proteins.

Example usage 

```bash
phold proteins-predict -i phage_proteins.faa -o phold_proteins_predict_output 
```

```bash
Usage: phold proteins-predict [OPTIONS]

  Runs ProstT5 on a multiFASTA input - GPU recommended

Options:
  -h, --help                     Show this message and exit.
  -V, --version                  Show the version and exit.
  -i, --input PATH               Path to input multiFASTA file  [required]
  -o, --output PATH              Output directory   [default: output_phold]
  -t, --threads INTEGER          Number of threads  [default: 1]
  -p, --prefix TEXT              Prefix for output files  [default: phold]
  -d, --database TEXT            Specific path to installed phold database
  -f, --force                    Force overwrites the output directory
  --batch_size INTEGER           batch size for ProstT5. 1 is usually fastest.
                                 [default: 1]
  --cpu                          Use cpus only.
  --omit_probs                   Do not output 3Di probabilities from ProstT5
  --finetune                     Use finetuned ProstT5 model (PhrostT5).
                                 Experimental and not recommended for now
  --finetune_path TEXT           Path to finetuned model weights
  --save_per_residue_embeddings  Save the ProstT5 embeddings per resuide in a
                                 h5 file
  --save_per_protein_embeddings  Save the ProstT5 embeddings as means per
                                 protein in a h5 file
```

### `phold proteins-compare`

Identical to `phold compare`, but instead takes a FASTA input file of amino acid protein sequences. Useful for bulk annotation of phage proteins.

Example usage 

```bash
phold proteins-compare -i phage_proteins.faa --predictions_dir phold_proteins_predict_output -o phold_proteins_compare_output  -t 8
```

```bash
Usage: phold proteins-compare [OPTIONS]

  Runs Foldseek vs phold db on proteins input

Options:
  -h, --help                Show this message and exit.
  -V, --version             Show the version and exit.
  -i, --input PATH          Path to input file in multiFASTA format
                            [required]
  --predictions_dir PATH    Path to output directory from phold proteins-
                            predict
  --structures              Use if you have .pdb or .cif file structures for
                            the input proteins (e.g. with AF2/Colabfold) in a
                            directory that you specify with --structure_dir
  --structure_dir PATH      Path to directory with .pdb or .cif file
                            structures. The CDS IDs need to be in the name of
                            the file
  --filter_structures       Flag that creates a copy of the .pdb or .cif files
                            structures with matching record IDs found in the
                            input GenBank file. Helpful if you have a
                            directory with lots of .pdb files and want to
                            annotate only e.g. 1 phage.
  -o, --output PATH         Output directory   [default: output_phold]
  -t, --threads INTEGER     Number of threads  [default: 1]
  -p, --prefix TEXT         Prefix for output files  [default: phold]
  -d, --database TEXT       Specific path to installed phold database
  -f, --force               Force overwrites the output directory
  -e, --evalue FLOAT        Evalue threshold for Foldseek  [default: 1e-3]
  -s, --sensitivity FLOAT   Sensitivity parameter for foldseek  [default: 9.5]
  --keep_tmp_files          Keep temporary intermediate files, particularly
                            the large foldseek_results.tsv of all Foldseek
                            hits
  --card_vfdb_evalue FLOAT  Stricter Evalue threshold for Foldseek CARD and
                            VFDB hits  [default: 1e-10]
  --separate                Output separate GenBank files for each contig
  --max_seqs INTEGER        Maximum results per query sequence allowed to pass
                            the prefilter. You may want to reduce this to save
                            disk space for enormous datasets  [default: 10000]
  --only_representatives    Foldseek search only against the cluster
                            representatives (i.e. turn off --cluster-search 1
                            Foldseek parameter)
  --ultra_sensitive         Runs phold with maximum sensitivity by skipping
                            Foldseek prefilter. Not recommended for large
                            datasets.
```

### `phold remote`

This command queries the Foldseek webserver to predict the 3Di sequence instead of running ProstT5 locally, followed by a local Foldseek search against the Phold database locally. This is recommended for users with extremely low compute (such as an old laptop) or who can't get ProstT5 to run on their machine.

Example usage 

```bash
phold remote -i pharokka.gbk  -o phold_remote_output  -t 8
```

```bash
Usage: phold remote [OPTIONS]

  Uses Foldseek API to run ProstT5 then Foldseek locally

Options:
  -h, --help                Show this message and exit.
  -V, --version             Show the version and exit.
  -i, --input PATH          Path to input file in Genbank format or nucleotide
                            FASTA format  [required]
  -o, --output PATH         Output directory   [default: output_phold]
  -t, --threads INTEGER     Number of threads  [default: 1]
  -p, --prefix TEXT         Prefix for output files  [default: phold]
  -d, --database TEXT       Specific path to installed phold database
  -f, --force               Force overwrites the output directory
  -e, --evalue FLOAT        Evalue threshold for Foldseek  [default: 1e-3]
  -s, --sensitivity FLOAT   Sensitivity parameter for foldseek  [default: 9.5]
  --keep_tmp_files          Keep temporary intermediate files, particularly
                            the large foldseek_results.tsv of all Foldseek
                            hits
  --card_vfdb_evalue FLOAT  Stricter Evalue threshold for Foldseek CARD and
                            VFDB hits  [default: 1e-10]
  --separate                Output separate GenBank files for each contig
  --max_seqs INTEGER        Maximum results per query sequence allowed to pass
                            the prefilter. You may want to reduce this to save
                            disk space for enormous datasets  [default: 10000]
  --only_representatives    Foldseek search only against the cluster
                            representatives (i.e. turn off --cluster-search 1
                            Foldseek parameter)
  --ultra_sensitive         Runs phold with maximum sensitivity by skipping
                            Foldseek prefilter. Not recommended for large
                            datasets.
```



### `phold createdb`

This in an auxillary command that allows you to create a Foldseek compatible database from AA and 3Di protein sequences (such as those created by `phold predict`). 

Example usage 

```bash
phold createdb --fasta_aa phold_aa.fasta  --fasta_3di phold_3di.fasta -o my_foldseek_db  
```


```bash
Usage: phold createdb [OPTIONS]

  Creates foldseek DB from AA FASTA and 3Di FASTA input files

Options:
  -h, --help             Show this message and exit.
  -V, --version          Show the version and exit.
  --fasta_aa PATH        Path to input Amino Acid FASTA file of proteins
                         [required]
  --fasta_3di PATH       Path to input 3Di FASTA file of proteins  [required]
  -o, --output PATH      Output directory   [default:
                         output_phold_foldseek_db]
  -t, --threads INTEGER  Number of threads to use with Foldseek  [default: 1]
  -p, --prefix TEXT      Prefix for Foldseek database  [default:
                         phold_foldseek_db]
  -f, --force            Force overwrites the output directory
```

### `phold plot`

This in an auxillary command that allows you to create Circos plots for your phage(s) with [pyCirclize](https://github.com/moshi4/pyCirclize). It requires only the `phold` Genbank file as its input and will output .png and .svg format files.

If you have annotated more than 1 contig with `phold`, `phold plot` will automatically plot them all in separate files. The contig ids will be the prefix of the output plot file names.

Example usage 

```bash
phold plot -i phold.gbk -o phold_plots
```


```bash
Usage: phold plot [OPTIONS]

  Creates Phold Circular Genome Plots

Options:
  -h, --help                      Show this message and exit.
  -V, --version                   Show the version and exit.
  -i, --input PATH                Path to input file in Genbank format (in the
                                  phold output directory)  [required]
  -o, --output PATH               Output directory to store phold plots
                                  [default: phold_plots]
  -p, --prefix TEXT               Prefix for output files. Needs to match what
                                  phold was run with.  [default: phold]
  -f, --force                     Force overwrites the output directory
  -a, --all                       Plot every contig.
  -t, --plot_title TEXT           Plot title. Only applies if --all is not
                                  specified. Will default to the phage\'s
                                  contig id.
  --label_hypotheticals           Flag to label hypothetical or unknown
                                  proteins. By default these are not labelled
  --remove_other_features_labels  Flag to remove labels for
                                  tRNA/tmRNA/CRISPRs. By default these are
                                  labelled.  They will still be plotted in
                                  black
  --title_size FLOAT              Controls title size. Must be an integer.
                                  Defaults to 20
  --label_size INTEGER            Controls annotation label size. Must be an
                                  integer. Defaults to 8
  --interval INTEGER              Axis tick interval. Must be an integer. Must
                                  be an integer. Defaults to 5000.
  --truncate INTEGER              Number of characters to include in annoation
                                  labels before truncation with ellipsis.
                                  Must be an integer. Defaults to 20.
  --dpi INTEGER                   Resultion \(dots per inch\). Must be an
                                  integer. Defaults to 600.
  --annotations FLOAT             Controls the proporition of annotations
                                  labelled. Must be a proportion between 0 and
                                  1 inclusive.  0 = no annotations, 0.5 = half
                                  of the annotations, 1 = all annotations.
                                  Defaults to 1. Chosen in order of CDS size.
  --label_ids TEXT                Text file with list of CDS IDs \(from gff
                                  file\) that are guaranteed to be labelled.
```